flink-runtime/src/main/java/org/apache/flink/streaming/api/graph/StreamingJobGraphGenerator.java

```diff
@@ -1678,10 +1678,27 @@ public static ResultPartitionType determineUndefinedResultPartitionType(
     public static boolean isChainable(StreamEdge edge, StreamGraph streamGraph) {
         StreamNode downStreamVertex = streamGraph.getTargetVertex(edge);
 
-        return downStreamVertex.getInEdges().size() == 1 && isChainableInput(edge, streamGraph);
+        return downStreamVertex.getInEdges().size() == 1
+                && isChainableInput(edge, streamGraph)
+                && checkAndSetParallelismForChainableNodes(edge, streamGraph);
```

Comment:
It's a bit weird to update node parallelism/maxParallelism in a method named `isChainable()`.

Reply:
How about to do the update when doing `ForwardPartitioner` conversion?